Sources/core/objects/Projectiles.cpp

```diff
@@ -0,0 +1,233 @@
+#include "Projectiles.h"
+#include "../Helpers.h"
+
+std::ostream& operator<<(std::stringstream& file,const std::vector<Dir>& s)
+{
+    file << " " << s.size() << " ";
+    for (auto it : s)
+        file << " " << it << " ";
+    return file;
+}
+
+std::istream& operator>>(std::stringstream& file,std::vector<Dir>& s)
+{
+    size_t size;
+    file >> size;
+    s.reserve(size);
+
+    unsigned int local_id;
+    for (size_t i = 0; i < size; ++i)
+    {
+        file >> local_id;
+        s.push_back(local_id);
+    }
+    return file;
+}
+
+unsigned int hash(const std::vector<Dir>& content)
+{
+    unsigned int retval = 0;
+    int i = 1;
+    for (auto it = content.begin(); it != content.end(); ++it, ++i)
+    {
+        retval += *it * i;
+    }
+    return retval;
+}
+
+
+Projectile::Projectile(size_t id) : IMovable(id)
+{
+    SetPassable(D_ALL, Passable::EMPTY);
+    damage = 0;
+    SetSprite("icons/projectiles.dmi");
+    v_level = 5;
+    current_step_ = 0;
+    set_target_ = 0;
+    tickSpeed = 1;
+    pixSpeed = 4;
+    reached_target = 0;
+}
+void Projectile::Process()
+{
+   if (set_target_ == 1)
+   {
+       Dir step =  movement_[current_step_];
+       Rotate(step);
+       if(!CheckPassable())
+       {
+           return;
+       }
+       else if (reached_target)
+       {
+           MainMove();
+           return;
+       }
+       else if (current_step_ < movement_.size()-1 && !reached_target)
+       {
+           current_step_++;
+       }
+       else if(!reached_target && current_step_ == movement_.size()-1)
+       {
+           current_step_ =0;
+           Dir tmp = movement_[1];
+           movement_.clear();
+           movement_.push_back(tmp);
+           reached_target = 1;
+       }    
+       MainMove();
+   }
```

Comment:
Well that's because if it reaches the shooting target tile then It just goes in the main direction of shooting, it's hard to explain but maybe I can show you if that's better.